src/java.base/share/classes/jdk/internal/foreign/StringSupport.java

```diff
@@ -40,131 +42,207 @@
+                for (int j = 0; j < Long.BYTES; j++) {
+                    if (SCOPED_MEMORY_ACCESS.getByte(segment.sessionImpl(), segment.unsafeGetBase(), segment.unsafeGetOffset() + offset + j) == 0) {
+                        return requireWithinArraySize(offset + j - fromOffset, segment, fromOffset, toOffset);
+                    }
+                }
+            }
+        }
+        // Handle the tail
+        for (; offset < toOffset; offset++) {
+            byte val = SCOPED_MEMORY_ACCESS.getByte(segment.sessionImpl(), segment.unsafeGetBase(), segment.unsafeGetOffset() + offset);
+            if (val == 0) {
+                return requireWithinArraySize(offset - fromOffset, segment, fromOffset, toOffset);
             }
         }
+        throw nullNotFound(segment, fromOffset, toOffset);
+    }
 
-        // We are now on a long-aligned boundary so this is the "body"
-        int bodyCount = bodyCount(segment.byteSize() - start - headCount);
-
-        for (; offset < bodyCount; offset += Long.BYTES) {
-            // We know we are `long` aligned so, we can save on alignment checking here
-            long curr = segment.get(JAVA_LONG_UNALIGNED, start + offset);
-            // Is this a candidate?
-            if (mightContainZeroByte(curr)) {
-                for (int j = 0; j < 8; j++) {
-                    if (segment.get(JAVA_BYTE, start + offset + j) == 0) {
-                        return offset + j;
+    @ForceInline
+    public static int strlenShort(final AbstractMemorySegmentImpl segment,
+                                  final long fromOffset,
+                                  final long toOffset) {
+        final long length = toOffset - fromOffset;
+        segment.checkBounds(fromOffset, length);
+        if (length == 0) {
+            segment.scope.checkValidState();
+            throw stringTooLarge(segment, fromOffset, toOffset);
+        }
+        final long longBytes = length & LONG_MASK;
+        final long longLimit = fromOffset + longBytes;
+        long offset = fromOffset;
+        for (; offset < longLimit; offset += Long.BYTES) {
+            long val = SCOPED_MEMORY_ACCESS.getLongUnaligned(segment.sessionImpl(), segment.unsafeGetBase(), segment.unsafeGetOffset() + offset, !Architecture.isLittleEndian());
+            if (mightContainZeroShort(val)) {
+                for (int j = 0; j < Long.BYTES; j += Short.BYTES) {
+                    if (SCOPED_MEMORY_ACCESS.getShortUnaligned(segment.sessionImpl(), segment.unsafeGetBase(), segment.unsafeGetOffset() + offset + j, !Architecture.isLittleEndian()) == 0) {
+                        return requireWithinArraySize(offset + j - fromOffset, segment, fromOffset, toOffset);
                     }
                 }
             }
         }
+        // Handle the tail
+        // Prevent over scanning as we step by 2
+        final long endScan = toOffset & ~1; // The last bit is zero
+        for (; offset < endScan; offset += Short.BYTES) {
```

Comment:
Question: do we care about finding unaligned sequences of two consecutive zero bytes? Because I don't think this method can detect those. E.g. if the first zero byte in a terminator sequence starts at an odd offset I don't think we can recognize the terminator.

This is a bit of a weird side-effect, because now this method no longer behaves like a plain byte loop where we look for two consecutive bytes. But, I suppose, in a way it also makes sense: if I want to read a C string whose characters are 2-bytes each, then I definitively need to scan the segment (at given offset) two bytes at a time. A terminator at an "odd offset" from the start of the scan is not good. I believe the javadoc is a bit "light" on how this terminator is actually found :-)

Reply:
Even the fact that the last byte is ignored (if length is odd), or that up to 3 bytes might be ignored in the `strlenInt` might not be obvious. Maybe we should file a followup PR for improving the javadoc?

Reply:
https://bugs.openjdk.org/browse/JDK-8345292